src/reducers/custom-wildcard-field.ts

```diff
@@ -0,0 +1,92 @@
+
+import {Action} from '../actions';
+import {CUSTOM_WILDCARD_ADD, CUSTOM_WILDCARD_ADD_FIELD, CUSTOM_WILDCARD_MODIFY_TITLE,
+        CUSTOM_WILDCARD_REMOVE, CUSTOM_WILDCARD_REMOVE_FIELD} from '../actions/custom-wildcard-field';
+import {CustomWildcardField} from '../models/custom-wildcard-field';
+import {insertItemToArray, modifyItemInArray, removeItemFromArray} from './util';
+
+function modifyFieldsProperty(fields: string[]) {
+  return (customWildcardField: CustomWildcardField) => {
+    return {
+      ...customWildcardField,
+      fields
+    };
+  };
+}
+
+export function customWildcardFieldReducer(
+  customWildcardFields: Readonly<CustomWildcardField[]> = [],
+  action: Action
+): CustomWildcardField[] {
+  switch (action.type) {
+    case CUSTOM_WILDCARD_ADD: {
+      const {fields, type} = action.payload;
+
+      let index = action.payload.index;
+      if (!index) {
+        index = customWildcardFields.length;
+      }
+
+      return insertItemToArray(customWildcardFields, index, {
+        fields,
+        type,
+        title: null
+      });
+    }
+
+    case CUSTOM_WILDCARD_REMOVE: {
+      const {index} = action.payload;
+      return removeItemFromArray(customWildcardFields, index).array;
+    }
+
+    case CUSTOM_WILDCARD_ADD_FIELD: {
+      const {index, fields} = action.payload;
+
+      const originalFields: string[] = customWildcardFields[index].fields;
+      const originalFieldsIndex = originalFields.reduce((dict, field) => {
+        dict[field] = true;
+        return dict;
+      }, {});
+
+      let updatedFields: string[];
```

Comment:
- I would name this `addedFields`

- I don't understand why you need to separate the logic for `n>1` and `n=1`.   Using a functional style, you can write this in one-line-ish. 

- There is another problem with this logic for immutability at the end.  What is it?  (Think for another 5 mins and ask me if you don't know.)

Reply:
Need to fix the n=0 corner case (add a duplicate field -- nothing changes, don't return a new array)